[1단계 - 콘솔 기반 로또 게임] 세오(진솔) 미션 제출합니다.

jin123457 opened the pull request:
안녕하세요, 온스타😊

우선 늦게 PR 본문을 작성하게 된 점 진심으로 죄송합니다.
그럼에도 불구하고 양해해 주셔서 정말 감사드립니다!

이번 1단계 로또 미션을 진행하면서 부족한 점이 많겠지만, 온스타의 세심한 피드백을 통해 많이 배우고 성장하고 싶습니다.
혹시 미흡한 부분이 있다면 언제든지 말씀해 주세요.

앞으로의 과정에서도 잘 부탁드립니다.

감사합니다! 🙏😊
## 폴더 구조
```
📦javascript-lotto
 ┣ 📂__tests__
 ┃ ┣ 📜Lotto.test.js // 로또 객체 테스트 파일
 ┃ ┣ 📜LottoStatistics.test.js // 로또 통계 객체 테스트 파일
 ┃ ┣ 📜Machine.test.js // 로또 기계 객체 테스트 파일
 ┃ ┗ 📜validation.test.js // validation 테스트 파일
 ┣ 📂src
 ┃ ┣ 📂constants
 ┃ ┃ ┣ 📜config.js // 설정에 관련한 상수 파일
 ┃ ┃ ┗ 📜message.js // error, input, output에 관련한 상수 파일
 ┃ ┣ 📂controller
 ┃ ┃ ┗ 📜Controller.js
 ┃ ┣ 📂domain
 ┃ ┃ ┣ 📂model
 ┃ ┃ ┃ ┣ 📜Lotto.js // 로또 객체 파일
 ┃ ┃ ┃ ┣ 📜LottoStatistics.js // 로또 통계 객체 파일
 ┃ ┃ ┃ ┗ 📜Machine.js // 로또 기계 객체 파일
 ┃ ┃ ┗ 📜validation.js // 검증 파일
 ┃ ┣ 📂utils
 ┃ ┃ ┗ 📜pickNumberInList.js // 랜덤 파일
 ┃ ┣ 📂view
 ┃ ┃ ┣ 📜InputView.js // 입력 파일
 ┃ ┃ ┣ 📜OutputView.js // 출력 파일
 ┃ ┃ ┗ 📜readLineAsync.js
 ┃ ┣ 📜step1-index.js
 ┃ ┗ 📜step2-index.js
```
## 학습 목표
이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.
- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트  
<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->
- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [ ] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: 

## 리뷰 요청 & 논의하고 싶은 내용  
<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.  
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점
<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? --> 
💭 **이번 미션에서 TDD에 대해 가장 많이 고민했습니다.**

    처음 해보는 TDD라 아직 익숙하지 않아서 시간이 많이 걸렸고, 어려운 점도 많았습니다.
    특히 TDD 사이클을 돌면서 테스트 코드를 작성하다 보니,
    마지막에 validation 단계에서 try-catch문을 사용할 때 문제가 생겼습니다.

    저와 페어가 구현하고 싶은 방식대로 진행하려면 모든 validation을 변경해야 하는 상황이었거든요.
    결국 저희는 각 함수에서 try-catch문을 사용해 문제를 해결했지만,
    지금 생각해 보면 재활용 가능한 함수를 하나 만들어 처리했더라면 더 좋지 않았을까 하는 아쉬움이 듭니다.

    사실 저는 TDD의 가장 큰 장점이 빠른 피드백을 받을 수 있다는 점이라고 생각했는데요.
    막상 진행해 보니 설계가 완벽하지 않으면 어느 정도 구현이 진행되기 전까지 문제를 알아차리기 어려운 부분도 있다는 느낌을 받았습니다.
    제가 이렇게 이해한 것이 맞을까요? 혹시 제가 TDD를 잘못 이해하고 있는 건 아닌지 혹은 잘 못 사용한거지 궁금합니다.

💭 **이번 미션에서는클래스의 역할과 책임을 적절하게 분리하는 방법을 고민하는 게 참 어려웠습니다.**

    처음엔 "작게 나누자"는 원칙을 생각하면서 클래스를 최대한 분리했는데요.
    그러다 보니 로또 객체가 오히려 너무 하는 일이 없어서
    "굳이 이걸 클래스로 나눴어야 했을까?" 하는 생각이 들기도 했고,
    반대로 더 많은 책임을 줬어야 했나? 하는 고민도 생겼습니다.
    
    반면에 로또 통계 객체에서는 오히려 너무 많은 역할을 하게 돼서
    "이게 정말 모델 역할을 제대로 하고 있는 걸까?" 하는 의문이 들기도 했습니다.
    
    이번 경험을 통해 클래스의 역할과 책임을 어디까지 부여해야 하는지,
    그리고 어느 시점에서 클래스를 분리하는 게 좋은지에 대해 더 깊이 고민해 봐야겠다는 생각이 들었었습니다.

### 2) 이번 리뷰를 통해 논의하고 싶은 부분  
<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->
- **TDD를 진행할 때 커밋 기준을 아직 잘 모르겠습니다.**
  
      TDD를 사용할 때 커밋 단위 기준을 논의해보고 싶습니다.
      저와 제 페어는 테스트 코드와 구현 코드를 분리해서 커밋했는데요.
      그 이유는 나중에 수정해야 하는 상황이 생겼을 때, 조금 더 쉽게 이전 상태로 돌아갈 수 있지 않을까 하는 생각에서였어요.
    
      하지만 이번 프로젝트에서는 커밋이 너무 많아지면서
      '너무 많으면 오히려 나중에 보기 어려운 게 아닐까?' 하는 고민도 생겼습니다.
    
      온스타는 이 부분에 대해 어떻게 생각하시나요?
      혹시 더 좋은 커밋 기준이 있다면 알려주시면 정말 감사하겠습니다! 😊

- **한눈에 의도를 파악할 수 있는 코드에 대해 고민했습니다.**

      복권 일치 개수를 파악하는 과정에서 페어와 함께 여러 가지 방법을 생각해 봤는데요.
    
      1. filter를 사용해 순회하면서 includes를 이용해 일치하는 개수를 찾는 방법
      2. 복권 숫자의 길이 + 당첨 숫자의 길이 - new Set(복권 숫자,당첨 숫자).size를 활용하는 방법
      이 두 가지 방법 중에서 어떤 것이 의도를 더 쉽게 전달할 수 있을지 고민이 되었습니다.
    
      온스타는 한눈에 의도를 파악할 수 있는 코드란 어떤 코드라고 생각하시나요?
      혹시 이 상황에서 더 좋은 접근 방식이 있다면 알려주시면 감사하겠습니다! 😊
    예시)
    ``` js
    /*
    machineLotto = [1,2,3,4,5,6];
    winningLotto = [1,2,3,4,5,7];
    */

    // 1안
    matchSameCount(machineLotto, winningLotto) {
      return machineLotto.filter((number) => winningLotto.includes(number)).length;
    }
    
    // 2안
    matchSameCount(machineLotto, winningLotto) {
      return (machineLotto.length + winningLotto.length) - new Set([...machineLotto,...winningLotto]).size;
    }
    ```
---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계
- [ ] TDD를 활용해 기능을 구현하는 과정에서 적절한 테스트 우선 접근 방식을 적용했는가? 단위 테스트의 커버리지는 충분한가?
- [ ] 도메인과 UI의 관심사를 분리하여 적절한 모듈화가 이루어졌는가? 하나의 객체나 모듈이 너무 많은 책임을 가지고 있지는 않은가?
- [ ] 객체의 프로퍼티를 직접 조작하기보다 메시지를 던지고 있는가?
- [ ] 불필요한 클래스를 사용하지 않고, 함수를 적극적으로 활용하여 JavaScript다운 방식으로 로직을 구현했는가?

### 2단계 
- [ ] 도메인 로직에 불필요한 영향을 주지 않고 UI 변경에 대응했는가?
- [ ] DOM 조작과 이벤트 활용을 JavaScript의 개념에 맞게 이해하고 적절하게 적용했는가? 
- [ ] 웹 표준을 준수하는 마크업을 활용하며, 스타일 작성에 일관성이 있는가?
</details>

kaori-killer committed:
feat: 숫자 6개를 가진 로또 1장을 발행

kaori-killer committed:
test: 숫자 6개를 가진 로또 1장을 발행한다

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 로또 번호의 숫자 범위는 1 ~ 45만 가능

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 로또의 번호의 숫자 범위는 1 미만 혹은 45 초과하면 에러가 발생

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 로또의 번호의 숫자 범위는 1 미만 혹은 45 초과하면 에러가 발생

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 로또 번호를 오름차순으로 정렬

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 로또 번호를 오름차순으로 정렬

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 구입 금액만큼 로또를 발행

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 구입 금액만큼 로또를 발행

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 최소/최대 범위를 가지고 랜덤 숫자를 주어진 길이만큼 생성 후 반환

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
style: 우테코 미션에 맞게 Lint 설정

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 한 장의 로또를 발행한다

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
refactor: 로또 번호의 최대/최소 값을 상수화

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
fix: 랜덤 값 추출할 때 범위를 넘어간 수 제거

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: '사용자가 구매한 로또 번호와 당첨 번호의 일치 개수를 반환

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 사용자가 구매한 로또 번호와 당첨 번호의 일치 개수를 반환

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 사용자가 구매한 로또 번호와 당첨된 번호가 %s개 일치하면 해당 복권의 숫자를 +1 증가

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: test: 사용자가 구매한 로또 번호와 당첨된 번호가 %s개 일치하면 해당 복권의 숫자를 +1 증가

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 사용자가 구매한 로또 번호와 보너스 번호를 비교한 후 참/거짓 값을 반환

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 사용자가 구매한 로또 번호와 보너스 번호를 비교한 후 참/거짓 값을 반환

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 사용자가 구매한 로또 번호와 당첨된 번호가 5개 일치하고 보너스 번호도 일치할 경우, 해당 복권의 숫자를 +1 증가

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 사용자가 구매한 로또 번호와 당첨된 번호가 5개 일치하고 보너스 번호도 일치할 경우, 해당 복권의 숫자를 +1 증가

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
chore: Lint를 적용하지 않을 파일 추가

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 사용자가 구매한 로또 번호화 당첨 번호를 비교하여 수익률 계산

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 사용자가 구매한 로또 번호화 당첨 번호를 비교하여 수익률 계산

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 수익률은 소숫점 둘째짜리에서 반올림

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: feat: 수익률은 소숫점 둘째짜리에서 반올림

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
docs: 기능 명세서 업데이트

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 로또 구입 금액을 입력

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
chore: 파일을 불러올 때 js 확장자 추가

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 당첨 번호 입력

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 보너스 번호를 입력

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 재시작/종료 여부를 입력

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
fix: 랜덤한 숫자 6개를 추출할 때 중복되는 숫자가 없음

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
refactor: 입력받는 부분 각 함수로 분리

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 로또 구입 금액은 1,000원으로 나누어 떨어짐

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 로또 구입 금액은 1,000원으로 나누어 떨어짐

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 로또 구입 금액은 0원 이하일 경우, 에러를 발생

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 로또 구입 금액은 0원 이하일 경우, 에러를 발생

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
refactor: 빈 배열은 split 함수를 적용하지 않음

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 로또 번호의 숫자 범위 1 ~ 45만 가능

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: feat: 로또 번호의 숫자 범위 1 ~ 45만 가능

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
refactor: 중복해서 사용하는 유효성 검사 함수는 공용 파일로 이동

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
style: 유효성 검사 함수 이름 변경

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 로또 번호는 6자리로 유효성 검사

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 로또 번호는 6자리로 유효성 검사

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 중복된 숫자가 있으면 안되는 유효성 검사

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 중복된 숫자가 있으면 안되는 유효성 검사

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 보너스 번호의 숫자 범위는 1 ~ 45 유효성 검사

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 보너스 번호의 숫자 범위는 1 ~ 45 유효성 검사

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 보너스 번호와 당첨번호는 중복된 숫자가 있으면 안되는 유효성 검사

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
test: 보너스 번호와 당첨번호는 중복된 숫자가 있으면 안되는 유효성 검사

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 사용자 로또의 수량을 계산

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 로또의 수량을 출력

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 사용자가 구매한 로또 번호를 출력

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 당첨 통계를 출력

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
refactor: 당첨 통계 제목을 출력

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 총 수익률을 출력

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
fix: 한 개의 로또의 숫자를 가져옴

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
refactor: domain과 ui 파일 분리

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
refactor: 불필요한 함수 제거

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
refactor: 매직 넘버 상수화

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
refactor: 입력받을 때 메시지를 상수화

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
fix: export 내보내기 방식에 맞게 파일 받기

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
docs: README.md 파일 수정

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
refactor: 객체의 키 값을 상수화

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
refactor: 매직 넘버를 상수화

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
feat: 예외 발생한 부분부터 입력을 다시 받음

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
refactor:  함수 분리 및 상수화

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

kaori-killer committed:
fix: 테스트 코드와 메인 코드의 에러 메시지를 동일하게 함

Co-Authored-By: seo_dev <72060681+jin123457@users.noreply.github.com>

onschan commented on src/constants/config.js:
```diff
@@ -0,0 +1,22 @@
+  DECIMAL: 10,
+  ANSWER_NO: 'n',
+  ANSWER_YES: 'y',
+});
```
상수를 관리하는 객체의 depth가 조금 깊어보여요.

도메인에 필요한 상수와 UI에 필요한 상수를 분리해주는 방법도 좋을 것 같아요

onschan commented on __tests__/Lotto.test.js:
```diff
@@ -0,0 +1,32 @@
+
+  // then
+  expect(lotto.getNumbers()).toHaveLength(6);
+});
```
6개 미만, 초과 혹은 1 ~ 45를 벗어나는 예외 케이스에 대한 테스트 코드가 추가된다면 더 신뢰성있는 테스트가 될 것 같아요.

onschan commented on __tests__/Lotto.test.js:
describe로 어떤 테스트를 진행하는지 명확히 표현해보는건 어떨까요?

onschan commented on __tests__/LottoStatistics.test.js:
```diff
@@ -0,0 +1,57 @@
+  const isBonusNumber = true;  
+  lottoStatistics.determineRank(sameCount, isBonusNumber);
+
+  expect(lottoStatistics.getRankResult()[`${sameCount}개 일치, 보너스 볼 일치`].count).toBe(1);
```
key값이 노출될 텍스트 자체로 관리되고있네요. 어떤 코드인지 햇갈렸네요.
해당 key 값은 텍스트 대신 다른 값으로 관리되는 게 좋을 것 같긴하겠지만 
테스트 코드에 대해서만 말씀드리면 하드 코딩된 텍스트를 테스트 코드에서 이용하는 건 테스트의 신뢰도를 크게 낮출 수 있어요.

onschan commented on __tests__/validation.test.js:
전반적으로 신뢰성있고 잘 짜여진 테스트 코드로 보여요. 다만 텍스트를 그대로 노출하는 대신 validation에서 이용하는 ERROR 상수를 그대로 이용해주시는게 좋을 것 같아요

onschan commented on src/domain/model/Lotto.js:
```diff
@@ -0,0 +1,16 @@
+    this.#numbers = numbers;
+  }
+
+  getNumbers() {
+    return this.#numbers.sort((a, b) => a - b);
+  }
```
매 호출 시마다 정렬을 수행해서 성능적으로 불리할 수 있어보여요.
생성자에서 직접 정렬하고 getter의 역할만 수행하는게 좋을 것 같아요!

jin123457 replied:
저는 출력할 때만 정렬된 배열이 필요하고, 도메인 로직에서는 굳이 정렬 기능이 필요 없다고 생각했어요.
그런데 온스타의 피드백을 받고 나니, “getter에서 정렬을 하는 게 과연 맞는 걸까?” 하는 고민이 생겼습니다.

그래서 랜덤 값을 뽑아올 때 정렬을 하는 방식으로 수정하려고 하는데요.
다만, 성능적으로 어떻게 불리할 수 있는지에 대해서는 아직 잘 이해가 되지 않아서요.
혹시 자세히 설명해 주실 수 있을까요? 😊

onschan replied:
정렬을 출력에 대한 요구 사항으로 해석하고 분리한다는 시각은 너무 좋은 것 같아요. 저도 미처 놓쳤던 부분인데요.

성능적으로 불리하다고 말씀드렸던 부분은 해당 배열이 10000개 정도의 요소를 가지고 있다고 가정을 하고 생각을 해보면 좋을 것 같아요. getter는 해당 필드가 필요한 부분마다 지속적으로 사용될 수 있는 부분이에요. 정렬에 대해 아무리 좋은 알고리즘을 적용했다고 하더라도 결국 정렬을 매번 수행하는 부분은 배열을 모두 순회해야되기 때문에, 단 한번만 정렬 수행 후 이미 정렬된 값을 쓰는 상황과 비교하면 성능적으로 불리할 수 있어요.

jin123457 replied:
친절한 답변 감사합니다ㅎㅎ☺️
온스타 말씀대로 여러 곳에서 사용할 수 있다는 사실을 미쳐 생각 못했네요ㅠ

생성할 때 정렬을 하는 방식으로 하는 것이 확실히 성능적으로 좋은 것 같습니다!

리팩터링에 반영하도록 하겠습니다😤

onschan commented on src/domain/model/LottoStatistics.js:
```diff
@@ -0,0 +1,65 @@
+      '5개 일치': { count: 0, price: 1500000 },
+      '5개 일치, 보너스 볼 일치': { count: 0, price: 30000000 },
+      '6개 일치': { count: 0, price: 2000000000 },
+    };
```
객체의 key값이 UI에 종속되어 있어보여요. 이 부분은 테스트 코드에서도 보였는데 도메인 로직과 분리해보도록 고려해주세요 🤔

onschan commented on src/domain/model/LottoStatistics.js:
```diff
@@ -0,0 +1,65 @@
+    const ONE_TICKET = 1;
+    if (sameCount < CONFIG.MIN.RANK_COUNT) {
+      return;
+    }
```
sameCount를 활용한 비교 연산이 increaseCount 메서드 사용처에서 선행되는 게 어떨까요?
increaseCount 메서드에서 sameCount가 어떤 역할을 할지 예측하기 어려울 것 같아요 🤔

onschan commented on src/domain/model/Machine.js:
Lottom 라는 키워드가 포함되는 편이 더 명확할 것 같아요!

onschan reviewed (changes requested):
안녕하세요. 세오! 😊
미션 진행하시느라 수고 많으셨어요.

저도 리뷰가 예상보다 늦어져서 죄송합니다 🥲

---

### 질문에 대한 답변

> 사실 저는 TDD의 가장 큰 장점이 빠른 피드백을 받을 수 있다는 점이라고 생각했는데요.
막상 진행해 보니 설계가 완벽하지 않으면 어느 정도 구현이 진행되기 전까지 문제를 알아차리기 어려운 부분도 있다는 느낌을 받았습니다.
제가 이렇게 이해한 것이 맞을까요? 혹시 제가 TDD를 잘못 이해하고 있는 건 아닌지 혹은 잘 못 사용한거지 궁금합니다.

저도 현업에서 TDD를 사용하고는 있지 않지만, 비지니스 로직 구현에 있어서 TDD는 충분히 활용될만하다고 생각하는 편이에요. (예전엔 비관론자에 가까웠습니다 😅)

세오가 느끼신 고민과 그를 수정해갔던 방향도 정말 가치있는 순간들이라고 생각해요.

TDD를 FE단에서 사용하는 회사들의 사례를 보다보면 자동화와 설계를 정말 중요시 생각하는 것 같아요. 선행된 많은 고민과 개선 작업을 통해 상황에 맞는 최선의 설계를 구축하고, 그에 따라 TDD를 위한 자동화를 이루고 난 뒤, 적극적으로 도입하는 과정을 거친다면 TDD가 개발 속도 향상에도 도움을 준다고도 하고요. 

TDD에 대해 익숙하지 않은 지금 당장 최적의 길을 걸을 필요는 없고 조바심을 느끼실 필요도 없을 것 같아요. 😊
우테코 처럼 다양한 경험을 해볼 수 있고 몰입할 수 있는 환경에서 세오의 기준이 생기고 룰이 생긴다면 어색해보이는 것들이 다시 보이기 시작할게요.

> 처음엔 "작게 나누자"는 원칙을 생각하면서 클래스를 최대한 분리했는데요.
그러다 보니 로또 객체가 오히려 너무 하는 일이 없어서
"굳이 이걸 클래스로 나눴어야 했을까?" 하는 생각이 들기도 했고,
반대로 더 많은 책임을 줬어야 했나? 하는 고민도 생겼습니다.

충분히 할 수 있고 가치있는 고민이네요. Lotto 클래스가 고민의 포인트셨을거라고 생각하는데, 세오가 고민한 것 처럼 Lotto 클래스는 불필요했을 수 있어요. 단순히 값으로 관리해도 도메인을 풀어내는데 문제가 없었을테고요.

다만 Lotto 클래스를 만는건 아무런 가치가 없었을까? 라고 생각한다면 그렇지 않다고 생각합니다. 도메인을 객체 간의 관계로 풀어내는 과정은 가상의 시뮬레이션을 만들어내는 것과 비슷하다고 생각해요. 로또를 활용하는 시뮬레이션에서 로또 객체를 만들어낸다는 상황은 자연스러운 흐름일거에요. 이 작은 로또 객체는 더 복잡한 로또를 만들어내는데 다시 사용될 수 있고요. (`class ComplexLotto extends Lotto`  같은 확장을 이용해서요.)

이번 미션의 요구 사항을 단순히 구현하는 과정에서는 Lotto라는 클래스 정의가 필요가 없었을 수도 있지만, 객체 간의 관계로 도메인을 표현하신 세오의 코드를 베이스해서는 Lotto 라는 클래스를 생성하시고 남겨두신 선택에 대해서도 동의합니다.

>하지만 이번 프로젝트에서는 커밋이 너무 많아지면서
'너무 많으면 오히려 나중에 보기 어려운 게 아닐까?' 하는 고민도 생겼습니다.
온스타는 이 부분에 대해 어떻게 생각하시나요?
혹시 더 좋은 커밋 기준이 있다면 알려주시면 정말 감사하겠습니다! 😊

기능 요구 사항과 그 구현 과정에 대해 설계를 이미 잘하셨고, 커밋도 잘 쪼개주셨고 다 좋습니다. 

작은 메서드 하나 만들고 커밋을 할 수도 있고 console.log 하나 지우고 커밋을 할 수도 있고, 내가 커밋을 할때 설명할 수 있는 목적이 있다면 그 자체의 커밋은 항상 유효합니다. 작업하는데 정말 불편함을 야기할 정도가 아니라면 커밋이 많아질 것에 대해서는 큰 걱정하실 필요는 없다고 생각해요.

저는 오히려 일을 하다보면 커밋을 너무 분리하지 않게 되어버리기도해서 종종 반성하고 있습니다.. 🥲
협업 과정에서 좀 더 고민해야될 부분은 PR의 작업 크기일 것 같아요.

> 온스타는 한눈에 의도를 파악할 수 있는 코드란 어떤 코드라고 생각하시나요?
혹시 이 상황에서 더 좋은 접근 방식이 있다면 알려주시면 감사하겠습니다! 😊
// 1안
matchSameCount(machineLotto, winningLotto) {
  return machineLotto.filter((number) => winningLotto.includes(number)).length;
}
// 2안
matchSameCount(machineLotto, winningLotto) {
  return (machineLotto.length + winningLotto.length) - new Set([...machineLotto,...winningLotto]).size;
}

1안이 더 적합한 코드라고 생각해요. matchSameCount라는 메서드 명에 맞게 “개수”에 집중을 한다면 2안 같은 코드가 나올 수도 있을거라고 생각은 하는데, “machineLotto와 winningLotto의 교집합”이라는 데이터 자체도 해당 도메인에서 충분히 중요한 데이터라고 생각하거든요. 어떤 숫자들로 당첨 됐는지 보여달라는 요구 사항도 충분히 생길수 있잖아요?

```tsx
//util
const getIntersectionArray(array1, array2) => {
    const set = new Set(array2);
    return array1.filter((item) => set.has(item));
}

// method
matchSameCount(machineLotto, winningLotto) {
  return getIntersectionArray(machineLotto, winningLotto).length;
}
```

위와 같은 코드를 사용한다면 n*m 대신 n의 시간 복잡도로 문제를 풀 수 있을 것 같은데, 굳이 1안에서 수정할만한 포인트는 느껴지지 않긴해요.

---
로또라는 도메인에 대해 잘 해석하고 잘 분리해주셨네요!
컨트롤러를 활용해서 도메인과 UI 분리에 대한 전략을 사용해주신 것도 좋은 전략이라고 생각합니다. 
테스트 코드도 고민하셨던 만큼 크게 문제점이 발견될 만한 사항도 없었어요.
전체적인 설계 방향과 구현 과정에 있어서 큰 이견은 없어요. 코멘트에 남긴 사항들만 잘 확인해주세요!

jin123457 committed:
style : .prettierrc 스타일 추가

jin123457 committed:
refactor : 도메인/UI 상수화 분리 및 호출 경로 수정

[삭제된 파일]
constants/constants.js : 도메인/UI의 상수를 서로 분리하기 위해 삭제되었습니다.

[생성된 파일]
domain/constants.js : 기존 CONFIG 안에 있었던 상수들을LOTTO_RULE, RANK_RULE,INITIAL_NUMBER 세개로 분리하여 내보내게끔 수정 및 생성하였습니다.
view/constants.js : 기존 CONFIG 안에 있었던 상수 중 UI에 관련된 상수들을 분리하여 수정 및 생성하였습니다.

[수정된 파일]
domain/validation.js : 기존 경로에서 호출 경로 수정 및 수정에 맞게 변경
domain/LottoStatistics.js : 기존 경로에서 호출 경로 수정 및 수정에 맞게 변경
domain/Machine.js : 기존 경로에서 호출 경로 수정 및 수정에 맞게 변경
view/InputView.js : 기존 경로에서 호출 경로 수정 및 수정에 맞게 변경
utils/pickNumberInList.js : 인자 값을 받아서 넘기는 방식에서 constants에서 LOTTO_RULE을 가지고 와서 생성하는 방안으로 수정

jin123457 committed:
test : 로또 생성 테스트 케이스 추가 및 그룹화

[수정된 파일]
Lotto.test.js :
* 숫자가 6개가 아닐 경우 에러를 발생 기능 추가
* 로또의 각 숫자 범위가 기준점보다 낮거나 높을시 에러 발생 기능 추가
* 로또의 숫자 중 중복됐을 경우 에러 발생 기능 추가
* 로또의 숫자 범위 그룹화
* 로또의 숫자의 개수 그룹화

jin123457 committed:
refactor : 도메인/UI 상수화 분리 및 호출 경로 수정

[수정된 파일]
controller/controller.js : 기존 경로에서 호출 경로 수정 및 수정에 맞게 변경

jin123457 committed:
test : 에러 문구 하드 코딩에서 상수화 수정

[수정된 파일]
test/validation.test.js : 에러 문구를 하드 코딩된 문자열에서 상수화된 에러 문구를 이용하게끔 변경

jin123457 committed:
rename : Machine에서 LottoMachine으로 파일, 클래스 명칭 수정

[수정된 파일]
test/Machine.test.js : Machine -> LottoMachine으로 파일 명칭 수정 및 호출 수정
controller/controller.js : Machine -> LottoMachine으로 호출 수정
domain/Machine.js : Machine -> LottoMachine으로 클래스 및 파일 명칭 수정

jin123457 committed:
test : Machine 생성 모델명 수정

jin123457 committed:
refactor : rank의 key값 반환할  텍스트 -> 서수로 수정

[생성된 파일]
utils/rankToString.js : rank 서수를 인자로 받아 출력 형식에 맞게 변환 후 반환

[수정된 파일]
test/LottoStatistics.test.js : rank의 key 값이 변경되면서 호출 방식 수정
constants/message.js : rankResult의 rank 출력 예시롤 변환 함수 추가
controller/controller.js : rankResult가 뒤부터 순회할 수 있도록 수
domain/constants.js : ONE_TICKET 값 상수화 추가
domain/LottoStatisics.js : rank의 키 값 텍스트에서 서수로 가독성 좋게 수정 및 determinRank함수 수정

jin123457 committed:
fix : 결과 누적 버그 수정

[수정된 파일]
controller/Controller.js : 결과가 재시작을 할 때 누적되는 현상이 발견되어 누적이 안되게끔 수정

jin123457 committed:
test : 로또 당첨 개수 테스트 케이스 수정

jin123457 commented:
안녕하세요 온스타😊
보내주신 피드백을 반영하여 리팩토링 진행했습니다.
제가 아직 많이 부족하다 보니, 말씀해 주신 요구사항을 제대로 이해하고 작업한 것이 맞는지 걱정이 됩니다.

이번 작업을 진행하면서 여러 가지를 고민하고 시도해 보았는데요.
혹시 부족한 부분이 있거나 리팩토링이 필요한 부분이 있다면 언제든 편하게 말씀해 주세요.
말씀해 주시면 바로 수정하도록 하겠습니다.

보내주신 리뷰 덕분에 많은 것을 배울 수 있었고, 여러 가지로 큰 도움이 되었습니다.
앞으로도 배울 점이 많을 것 같아 기대가 됩니다.

감사합니다. 앞으로도 잘 부탁드립니다! 😊

# 리팩토링 반영 사항
1. ♻️ 상수의 domain / UI 분리
    > 상수를 관리하는 객체의 depth가 조금 깊어보여요.
    도메인에 필요한 상수와 UI에 필요한 상수를 분리해주는 방법도 좋을 것 같아요
    
    constants를 기존에 constants 폴더 안에 두었는데 domain,view 안에 `constants.js` 파일을 생성하여 분리하는 방식으로 변경하였습니다.
    분리하는 과정에서 domain과 view에서 동시에 쓰이는 상수가 딱 하나가 있어서(재시작에 사용하는 Y/N) 제 생각에는 'view에 더 가깝지 않을까' 라는 고민 끝에 view 안에 넣어놓았습니다.
    이럴 경우 온스타는 어떻게 생각하시는지 궁금합니다💭
    
2.  🧪 테스트 케이스
    > 6개 미만, 초과 혹은 1 ~ 45를 벗어나는 예외 케이스에 대한 테스트 코드가 추가된다면 더 신뢰성있는 테스트가 될 것 같아요.
    
    > describe로 어떤 테스트를 진행하는지 명확히 표현해보는건 어떨까요?
    
    > 전반적으로 신뢰성있고 잘 짜여진 테스트 코드로 보여요. 다만 텍스트를 그대로 노출하는 대신 validation에서 이용하는 ERROR 상수를 그대로 이용해주시는게 좋을 것 같아요
    
     [test : 로또 생성 테스트 케이스 추가 및 그룹화](https://github.com/woowacourse/javascript-lotto/pull/361/commits/2089efe62762ace5163791ff9e1bf9ffa426bff8)
    * 6개 미만과 초과일 경우
    * 1 ~ 45를 벗어나는 경우
    * 로또 번호 안에 중복값이 있는 경우
    3가지 예외처리를 함과 동시에 LOTTO_RULE에 LENGTH, MIN_RANGE, MAX_RAGE를 이용하여 수정을 하여도 유동적으로 테스트도 같이 변할 수 있게끔 수정을 하였습니다.
    또한, ERROR 상수를 이용해 하드코딩이 아닌 상수의 문구를 이용할 수 있게 변경하였습니다.

3. 🤖 rank 가독성 증진 및 로직 변경
    > key값이 노출될 텍스트 자체로 관리되고있네요. 어떤 코드인지 햇갈렸네요.
    해당 key 값은 텍스트 대신 다른 값으로 관리되는 게 좋을 것 같긴하겠지만
    테스트 코드에 대해서만 말씀드리면 하드 코딩된 텍스트를 테스트 코드에서 이용하는 건 테스트의 신뢰도를 크게 낮출 수 있어요.
    
    > 객체의 key값이 UI에 종속되어 있어보여요. 이 부분은 테스트 코드에서도 보였는데 도메인 로직과 분리해보도록 고려해주세요 🤔
    
    [refactor : rank의 key값 반환할 텍스트 -> 서수로 수정](https://github.com/woowacourse/javascript-lotto/pull/361/commits/437c298790915da9607f1db2b24f160e8d41ef84)
    rank의 key값을 노출될 텍스트에서 서수로 변경하였습니다.
    서수는 가독성을 위해 일등을 제일 위에 두었고, 출력할 때 뒤집어서 출력하게끔 했습니다.
    또한 key 값을 UI 종속을 안 시키기 위해 rankToString 함수를 이용했습니다.
    테스트 코드에서 하드 코딩된 값을 없애여 신뢰도를 올리기 위해 아래와 같이 테스트를 수정했습니다.
    하지만, 결국 이것도 하드 코딩 아닌가 라는 생각에 함수를 따로 결정 짓는 것을 생성해야하나 라는 고민이 있습니다.
    온스타는 어떻게 생각하시나요?💭
    ``` js
    test.each([
      [3, false, 'fifth'],
      [4, false, 'fourth'],
      [6, false, 'first'],
      [5, false, 'third'],
      [5, true, 'second'],
    ])(
      '사용자가 구매한 로또 번호와 당첨된 번호가 %s개 일치하면 해당 복권의 숫자를 +1 증가시킨다.',
      (sameCount, bonus, prize) => {
        lottoStatistics.determineRank(sameCount, bonus);
    
        expect(lottoStatistics.getRankResult()[prize].count).toBe(ONE_TICKET);
      },
    );
    ```

jin123457 committed:
test : rankToString 테스트 추가

jin123457 committed:
refactor : 도메인에서 안쓰는 상수 제거

jin123457 committed:
test : getLottoQuantity 함수에 대한 테스트 케이스 추가

jin123457 committed:
test: restart 함수에 대한 테스트 케이스 추가

onschan commented on src/utils/rankToString.js:
```diff
@@ -0,0 +1,11 @@
+  if (prize === 'fifth') return PRIZE_MESSAGES.fifth;
+
+  return '';
+}
```
해당 함수가 PRIZE_MESSAGES[prize]의 역할을 그대로하고있는 것 같은데 꼭 필요할까? 생각이 드는데 어떻게 생각하시나요?

jin123457 replied:
지금 다시 확인해보니 정말 만들 필요가 없는 함수였네요,,,
제가 리팩터링을 하면서 너무 자신만의 생각에 매몰 되어있었던 것 같아요ㅠ
알려주셔서 감사합니다.🥹

onschan replied:
코드 작성 당시 본인의 기준과 시각이 다 있었던 부분이니 리뷰된 사항에 대해 자책할 필요는 없어요! 
저는 전체적으로 다시 보고있기때문에 보이는 부분들이 있는구요! 충분히 멋지게 잘 하고 계셔요 😋

onschan reviewed (changes requested):
안녕하세요. 세오!

리뷰 잘 반영해주시고 반영해주시면서 느끼신 고민들 잘 공유해주셔서 감사합니다 ☺️

>constants를 기존에 constants 폴더 안에 두었는데 domain,view 안에 `constants.js` 파일을 생성하여 분리하는 방식으로 변경하였습니다.
분리하는 과정에서 domain과 view에서 동시에 쓰이는 상수가 딱 하나가 있어서(재시작에 사용하는 Y/N) 제 생각에는 'view에 더 가깝지 않을까' 라는 고민 끝에 view 안에 넣어놓았습니다.
이럴 경우 온스타는 어떻게 생각하시는지 궁금합니다💭

domain과 view에 쓰이는 상수가 `ANSWER_YES`, `ANSWER_NO` 이 상수 말씀하신 것 같은데요. 왜 동시에 쓰인다고 생각하셨을까요?

```tsx
function validateRestart(lowerCaseInput) {
  if (lowerCaseInput !== ANSWER_YES && lowerCaseInput !== ANSWER_NO) {
    throw new Error(ERROR.RESTART.YES_OR_NO);
  }
}
```

이 유효성 검사 함수는 로또라는 도메인과 관련되어있을까요? 아니면 입.출력과 관련된 함수일까요? 상수도 성격에 따라 분리를 하신다면 유효성 검사도 성격에 따라 분리를 해보면 어떨까요?

>또한 key 값을 UI 종속을 안 시키기 위해 rankToString 함수를 이용했습니다.
테스트 코드에서 하드 코딩된 값을 없애여 신뢰도를 올리기 위해 아래와 같이 테스트를 수정했습니다.
하지만, 결국 이것도 하드 코딩 아닌가 라는 생각에 함수를 따로 결정 짓는 것을 생성해야하나 라는 고민이 있습니다.
온스타는 어떻게 생각하시나요?💭

제가 하드코딩이라고 표현드렸던 부분은 노출될 텍스트가 어떤 약속된 관리 방식없이 그대로 노출되고 있는 부분이 오타 등의 휴먼 에러나 요구 사항 변경 등에 대응이 어렵다는 부분이 있었기 때문이에요. 

약속된 데이터 형식으로 관리하고 그 데이터 형식에 따라 도메인 간의 관계를 그려내는 부분으로 수정이 되었다면 그걸로 충분할 것 같습니다. 다만 rankToString 함수의 위치나 역할에 대해 조금 더 고민해보시죠!

---
질문 사항에서 아직 의문이 남으시는 부분이 있는 것 같아서 한번 더 고민해보시고 편하게 리뷰 요청 다시 해주세요!

jin123457 committed:
refactor: 상수 ANSWER 위치 이동 및 관련 코드 수정

jin123457 committed:
refactor: rankToString 함수 제거 및 PRIZE_MESSAGES로 대체

jin123457 committed:
refactor: pickNumberInList 함수에 LOTTO_RULE 매개변수 추가

jin123457 committed:
refactor: getNumbers 메서드에서 정렬 제거 및 createLotto에서 정렬 추가

jin123457 commented:
안녕하세요 온스타👋
리팩터링 반영하고 반영 사항에 대해 정리 글 올립니다ㅎㅎ

## 리팩터링 반영 사항
> domain과 view에 쓰이는 상수가 ANSWER_YES, ANSWER_NO 이 상수 말씀하신 것 같은데요. 왜 동시에 쓰인다고 생각하셨을까요?
> 이 유효성 검사 함수는 로또라는 도메인과 관련되어있을까요? 아니면 입.출력과 관련된 함수일까요? 상수도 성격에 따라 분리를 하신다면 유효성 검사도 성격에 따라 분리를 해보면 어떨까요?
``` js
function validateRestart(lowerCaseInput) {
  if (lowerCaseInput !== ANSWER_YES && lowerCaseInput !== ANSWER_NO) {
    throw new Error(ERROR.RESTART.YES_OR_NO);
  }
}
```
해당 피드백을 반영하여 validation을 분리하려고 했으나, validation을 분리하지 않고 `ANSWER_YES`, `ANSWER_NO`를 domain으로 옮기는 것이 더 바람직하다고 판단했습니다.

이에 따라 `inputView`에서는 해당 상수를 사용하지 않고, 입력된 값을 그대로 반환한 후 `controller`에서 판단하도록 로직을 변경했습니다.
``` js
/* Before */
ANSWER_YES = 'y';
ANSWER_NO = 'n';

/* After */
ANSWER = Object.freeze({
  RE_START : 'y',
  END : 'N'
});
```
상수를 `view`에서 `domain/constants.js`로 이동하면서, 가독성을 더욱 높이기 위해 상수명을 수정했습니다. 😊

앞으로도 주신 피드백을 바탕으로 더욱 발전할 수 있도록 노력하겠습니다.
다시 한번 소중한 의견을 나눠 주셔서 감사해요 😊

onschan reviewed (approved):
수고많으셨어요 세오!
이번 단계는 여기서 머지하고 다음 단계에서 뵈어도 될 것 같아요!
다음 미션에서도 많이 고민해주시고 어떤 고민하셨는지 알려주세요 😊
